src/peer/db/statement-db.js

```diff
@@ -0,0 +1,168 @@
+// @flow
+
+const path = require('path')
+const Knex = require('knex')
+const pb = require('../../protobuf')
+
+import type { StatementMsg, SimpleStatementMsg, EnvelopeStatementMsg, CompoundStatementMsg } from '../../protobuf/types'
+
+const MIGRATIONS_DIR = path.join(__dirname, 'migrations')
+
+type StatementDBOptions = {
+  filename: string,
+}
+
+const DefaultOptions: StatementDBOptions = {
+  filename: ':memory:'
+}
+
+class StatementDB {
+  _db: Object
+  _migrated: boolean
+
+  constructor (options: StatementDBOptions = DefaultOptions) {
+    this._db = Knex({
+      client: 'sqlite3',
+      connection: {
+        filename: options.filename
+      },
+      useNullAsDefault: true
+    })
+    this._migrated = false
+  }
+
+  /**
+   * Accessor for the underlying sql DB (via knex query builder).
+   * Use this instead of directly accessing the `_db` instance,
+   * to ensure that the db migrations are run before accessing
+   * the db.
+   * @returns {*}
+   */
+  sqlDB (): Promise<Object> {
+    if (this._migrated) return Promise.resolve(this._db)
+
+    return this._db.migrate.latest({directory: MIGRATIONS_DIR})
+      .then(() => {
+        this._migrated = true
```

Comment:
Oh, damn.  This branch actually has a bug around this migration logic that was uncovered and fixed in the push branch (on https://github.com/mediachain/aleph/pull/147/commits/4bb01c385a328158f0ff07650ff5f88d56ab6755) - there's a race condition between checking the migrated flag and the migration completion, and it fails if you try to do two migrations concurrently.

I added a mutex on that commit, which makes it even more awkward... but it seems like some kind of lock is needed. 

I think there might be a cleaner way to do it with a condition variable (from the `locks` module that gives us the mutex) though... basically you could start the migration in the constructor, which will set the condition to true when the migration completes.  Then the `sqlDB` accessor can wait on the condition before returning the db. It should be a bit easier to follow.